File: js/arrays_strings/first_last_element_position_sorted_array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
const searchRange = (nums, target) => {
    left =0;
    right = nums.length -1;
    while(left <= right){
        mid = Math.floor((left+right)/2);
        if (nums[mid] == target){
            if (nums[mid] == 0 || nums[mid-1] != target ){
                return mid
            }
            right = mid -1
        }else if(nums[mid] > target ){
            right = mid -1
        }else{
            left = mid +1
        }
    }
};

nums = [10,11,11,11,14,15]; // [3,4]
target = 8;


// nums = [5,7,7,8,8,10]; // [3,4]
// target = 8;

// nums = [5,7,7,8,8,10]; // [-1,-1]
// target = 6

// nums = []; // [-1,-1]
// target = 0

console.log(searchRange(nums, target));

File: python/arrays_strings/first_last_element_position_sorted_array.py
from typing import List


class Solution:

    def getLeftPosition(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums)-1
        while(left <= right):
            mid = (left+right)//2
            if(nums[mid] == target):
                if(mid-1 >= 0 and nums[mid-1] != target or mid == 0):
                    return mid
                right = mid - 1
            elif(nums[mid] > target):
                right = mid-1
            else:
                left = mid+1
        return -1

    def getRightPosition(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums)-1
        while(left <= right):
            mid = (left+right)//2
            if(nums[mid]  == target):
                if(mid+1<len(nums) and nums[mid+1] != target or mid==len(nums)-1):
                    return mid
                left = mid + 1
            elif(nums[mid] > target):
                right = mid-1
            else:
                left = mid+1
        return -1

    def searchRange(self, nums: List[int], target: int) -> List[int]:
        left = self.getLeftPosition(nums, target)
        right = self.getRightPosition(nums, target)
        return [left, right]


s = Solution()
nums = [10,11,11,11,14,15]
target = 11
ans = s.searchRange(nums, target)
print(ans)